SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -142,17 +119,40 @@ void USpatialReceiver::OnAddComponent(Worker_AddComponentOp& Op)
 		GlobalStateManager->LinkAllExistingSingletonActors();
 		return;
 	case SpatialConstants::DEPLOYMENT_MAP_COMPONENT_ID:
- 		GlobalStateManager->ApplyDeploymentMapData(Op.data);
+		GlobalStateManager->ApplyDeploymentMapData(Op.data);
 		return;
 	case SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID:
- 		GlobalStateManager->ApplyStartupActorManagerData(Op.data);
+		GlobalStateManager->ApplyStartupActorManagerData(Op.data);
 		return;
-	default:
-		Data = MakeShared<improbable::DynamicComponent>(Op.data);
-		break;
 	}
 
-	PendingAddComponents.Emplace(Op.entity_id, Op.data.component_id, Data);
+	if (ClassInfoManager->IsSublevelComponent(Op.data.component_id))
+	{
+		return;
+	}
+
+	// If a client gains ownership over something it had already checked out, it will
+	// add component interest on the owner only data components, which will trigger an
+	// AddComponentOp, but it is not guaranteed to be inside a critical section.
+	if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(Op.entity_id))
+	{
+		if (ClassInfoManager->GetCategoryByComponentId(Op.data.component_id) == SCHEMA_OwnerOnly)
+		{
+			// We received owner only data, and we have the entity checked out already,
+			// so this happened as a result of adding component interest. Apply the data
+			// immediately instead of queuing it up (since there will be no AddEntityOp).
+			ApplyComponentData(Op.entity_id, Op.data, Channel);
```

Comment:
I think we should check that we aren't in a critical section when we try and do this.

Reply:
The AddComponentOp after the client sets its component interest may or may not come inside a critical section, but we need to apply it anyway. We can tell that it's cause by component interest because we already have the entity checked out.

Reply:
We shouldn't be applying this on the server though. Since at the moment OwnerOnly AddComponents will come through here and be applied.

Reply:
When would the server ever receive an AddComponentOp after it already has the entity checked out?